### PR TITLE
`Development`: Conditional Build Matrix

### PR DESCRIPTION
**Pull request by hhontheim**
## **TL;DR:** This PR allows better control over the images that are being built.

The newly added bash script `generate_build_matrix.sh` generates a conditional build matrix, based on changed directories.
It will only build images for those directories that contain a changed (or an added) `Dockerfile` or `RELEASE`-file.
It will then extract the version from the `RELEASE`-file and adds it to the matrix – this information will be used by the buildx-step which injects it into the Dockerfile.

Additionally, it generates two flags `build_base_images` and `build_bamboo_variants` that will also be used by the workflow to be able to skip steps completely.

## New `RELEASE`-file

Each directory now features an individual `RELEASE` file, where several options can be set.

### Versioning

Instead of hardcoded version numbers scattered around different files, there is now just a single line to be edited.
Each `RELEASE`-file contains a `VERSION` line which specifies the version the image will be based on.
Not only will this parameter affect the sources that will be downloaded, but this will also be used for tagging the final image.
The GitHub workflow will inject the specified version into the Dockerfile.

### Skipping builds

Each `RELEASE`-file contains a `BUILD` line, which can be set to `no` if a specific build should be skipped at all.

### Defining `PLATFORMS` for the Bamboo variants

The Bamboo variants for **Swift** and **C** contain a `PLATFORMS` line in their `RELEASE`-file.
This will be evaluated by the workflow to restrict variant builds to certain platforms.

## Locking the Swift version

The used Swift version is now locked to `focal` instead of using the `latest` tag.

The current `latest` build of the Swift image uses a compiler version of Swift 5.6.x.
Unfortunately, but actually rather strange, there is no current `arm64` image for Swift...
Instead of building our own Swift image as well, we use the focal tag, as this image uses Swift v5.5 which is available as a native `arm64` image.
We could also build the Swift image ourselves in the future – the corresponding Dockerfile for this is linked within a comment.

## Allowing Pull Requests

The new workflow allows test builds to be triggered by pull requests.
These builds are the same as the ones triggered by merges or commits to `main`.
However, these images will not be pushed to the container registry.

## Restrictions regarding caching

The Dockerfiles are now more optimized for caching the built Docker layers.
By moving commands that are less likely to change up in the Dockerfile, the build can be sped up by using cached versions of these layers.

However, there are some downsides while using GitHub Actions:

> GitHub will remove any cache entries that have not been accessed in over 7 days.
> There is no limit on the number of caches you can store, but the total size of all caches in a repository is limited to 10 GB.
> If you exceed this limit, GitHub will save your cache but will begin evicting caches until the total size is less than 10 GB.
> [See documentation here.](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows)

With the added images for GitLab and Jenkins by PR #4, we exceed the 10 GB limit.
Maybe we will find a workaround (other than moving each image to its separate repository), or maybe GitHub increases the limits in the future...

## Other minor changes

- Currently, building the Swift image will be skipped, as there are some errors building **SwiftLint**.
- The Bitbucket Dockerfile now supports builds of newer versions that use `opensearch` instead of `elasticsearch`, which caused crashes earlier.
- The tags will now be based on the repository's name. This enables easy builds on forks.
- Please check out the step `check-matrix` for runs of the workflow. Here one can verify the generated matrix more easily.

### CODE REVIEW

File: jira/RELEASE [hunk 1]
```diff
@@ -0,0 +1,2 @@
+BUILD=yes
+JIRA_VERSION=8.20.6
```
Comment (Mtze) [inline jira/RELEASE:2]: I guess this should be `VERSION` as well?

File: bitbucket/RELEASE [hunk 2]
```diff
@@ -0,0 +1,2 @@
+BUILD=yes
+BITBUCKET_VERSION=7.21.0
```
Comment (Mtze) [inline bitbucket/RELEASE:2]: Same as for jira

File: bamboo/RELEASE [hunk 3]
```diff
@@ -0,0 +1,2 @@
+BUILD=yes
+BAMBOO_VERSION=8.1.3
```
Comment (Mtze) [inline bamboo/RELEASE:2]: Same as for jira

**Comment by hhontheim** [comment]
**A hint regarding testing**: Since the images built in PRs are not released (by design), it's best to fork this repository and merge the feature branch into `main` there. This way, you'll have your own images to create containers from.

**Review by sleiss — COMMENTED** [review]
The changes look fine to me, I would however prefer if you could also document how one can build an image locally:
Previously, it was enough to just execute `docker build` for the desired Dockerfile, but now this fails since `VERSION` is no longer defined.

I think mentioning the use of `docker build .  --build-arg VERSION=...` should be sufficient for now, what do you think?

**Comment by sleiss** [comment]
Another question regarding the caching: You mentioned the issues with the evicted cache entries due to the time limitation as well as the space limitation.

Have you thought about inlining the cache into the Docker image (see [e.g. here](https://github.com/docker/build-push-action/blob/c5e6528d5ddefc82f682165021e05edf58044bce/docs/advanced/cache.md#inline-cache))?
You would then have to pull the (old version of the) image from the registry before building, but maybe this can solve the caching limitation?

File: generate_build_matrix.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
```
Comment (Mtze) [inline generate_build_matrix.sh:1]: Please add some comments/documentation to this file. I think it is not immediately obvious what the steps in this file do.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -56,42 +88,52 @@ jobs:
...
-          build-args: ${{ matrix.version }}
           context: ./${{ matrix.name }}
           file: ./${{ matrix.name }}/Dockerfile
+          build-args: VERSION=${{ matrix.version }}
           platforms: linux/arm64/v8,linux/amd64
-          push: true
-          tags: ghcr.io/ls1intum/artemis-${{ matrix.name }}:latest,ghcr.io/ls1intum/artemis-${{ matrix.name }}:${{ matrix.version }}
-          labels: "runnumber=${GITHUB_RUN_ID}"
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ghcr.io/${{ github.event.repository.owner.login }}/artemis-${{ matrix.name }}:latest,ghcr.io/${{ github.event.repository.owner.login }}/artemis-${{ matrix.name }}:${{ matrix.version }}
```
Comment (Mtze) [inline .github/workflows/build.yml:103]:
```suggestion
          tags: 
            - ghcr.io/${{ github.event.repository.owner.login }}/artemis-${{ matrix.name }}:latest
            - ghcr.io/${{ github.event.repository.owner.login }}/artemis-${{ matrix.name }}:${{ matrix.version }}
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -114,23 +156,28 @@ jobs:
...
-          build-args: ${{ matrix.version }}
           context: ./bamboo/${{ matrix.name }}
           file: ./bamboo/${{ matrix.name }}/Dockerfile
+          build-args: VERSION=${{ matrix.version }}
           platforms: ${{ matrix.platforms }}
-          push: true
-          tags: ghcr.io/ls1intum/artemis-bamboo:latest-${{ matrix.name }},ghcr.io/ls1intum/artemis-bamboo:${{ matrix.version }}-${{ matrix.name }}
-          labels: "runnumber=${GITHUB_RUN_ID}"
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ghcr.io/${{ github.event.repository.owner.login }}/artemis-bamboo:latest-${{ matrix.name }},ghcr.io/${{ github.event.repository.owner.login }}/artemis-bamboo:${{ matrix.version }}-${{ matrix.name }}
```
Comment (Mtze) [inline .github/workflows/build.yml:171]:
```suggestion
          tags: 
            - ghcr.io/${{ github.event.repository.owner.login }}/artemis-bamboo:latest-${{ matrix.name }}
            - ghcr.io/${{ github.event.repository.owner.login }}/artemis-bamboo:${{ matrix.version }}-${{ matrix.name }}
```

**Review by Mtze — CHANGES_REQUESTED** [review]
This is quite a dramatic change - I'm not sure if the introduced complexity by just building the images that changed is really necessary though. 
The atlassian versions change 2-4 times a year - and if they change, we update all the versions at once. So we have a minimal overhead there. I'm not sure how often the gitlab and jenkins images are update (cc @sleiss) that you add in the other PR. 

In any case - You have to improve the documentation for the new process. Please include at least: 
- How to add new Images to the build process 
- How to build the images locally 
- How the build matrix generation is working

I really appreciate the amount of thought you put in this PR!
